bpf/bpf_xdp.c

```diff
@@ -0,0 +1,196 @@
+ */
+
+/* TODOs XDP handling:
+ *
+ *   - Trace events to perf RB for sampling
+ *   - Tunnel secctx extraction and lookup
+ *   - Prog regeneration on the fly
+ *   - Htabs for often used prefixes (/16, /24)?
+ *   - Enable LPM when duming prefixes works form kernel
+ *   - Plugin framework
+ *   - Single lookup on prefix, revision
+ *   - XDP meta data transfer (#909)
+ */
+#include <node_config.h>
+#include <netdev_config.h>
+#include <filter_config.h>
+
+#include <bpf/api.h>
+
+#include <stdint.h>
+#include <stdio.h>
+
+#include <linux/bpf.h>
+#include <linux/if_ether.h>
+
+#include "lib/utils.h"
+#include "lib/common.h"
+#include "lib/maps.h"
+#include "lib/xdp.h"
+#include "lib/eps.h"
+#include "lib/events.h"
+
+#ifndef HAVE_LPM_MAP_TYPE
+# undef CIDR4_LPM_PREFILTER
+# undef CIDR6_LPM_PREFILTER
+#endif
+
+#ifdef CIDR4_FILTER
+struct bpf_elf_map __section_maps CIDR4_HMAP_NAME = {
+	.type		= BPF_MAP_TYPE_HASH,
+	.size_key	= sizeof(struct lpm_v4_key),
+	.size_value	= sizeof(struct lpm_val),
+	.flags		= BPF_F_NO_PREALLOC,
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CIDR4_HMAP_ELEMS,
+};
+
+#ifdef CIDR4_LPM_PREFILTER
+struct bpf_elf_map __section_maps CIDR4_LMAP_NAME = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct lpm_v4_key),
+	.size_value	= sizeof(struct lpm_val),
+	.flags		= BPF_F_NO_PREALLOC,
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CIDR4_LMAP_ELEMS,
+};
+#endif /* CIDR4_LPM_PREFILTER */
+#endif /* CIDR4_FILTER */
+
+#ifdef CIDR6_FILTER
+struct bpf_elf_map __section_maps CIDR6_HMAP_NAME = {
+	.type		= BPF_MAP_TYPE_HASH,
+	.size_key	= sizeof(struct lpm_v6_key),
+	.size_value	= sizeof(struct lpm_val),
+	.flags		= BPF_F_NO_PREALLOC,
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CIDR4_HMAP_ELEMS,
+};
+
+#ifdef CIDR6_LPM_PREFILTER
+struct bpf_elf_map __section_maps CIDR6_LMAP_NAME = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct lpm_v6_key),
+	.size_value	= sizeof(struct lpm_val),
+	.flags		= BPF_F_NO_PREALLOC,
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CIDR4_LMAP_ELEMS,
+};
+#endif /* CIDR6_LPM_PREFILTER */
+#endif /* CIDR6_FILTER */
+
+static __always_inline int check_v4_endpoint(struct xdp_md *xdp,
+					     struct iphdr *ipv4_hdr)
+{
+	if (lookup_ip4_endpoint(ipv4_hdr))
+		return XDP_PASS;
+
+	return XDP_DROP;
+}
+
+static __always_inline int check_v4(struct xdp_md *xdp)
+{
+	void *data_end = xdp_data_end(xdp);
+	void *data = xdp_data(xdp);
+	struct iphdr *ipv4_hdr = data + sizeof(struct ethhdr);
+	struct lpm_v4_key pfx __maybe_unused;
+
+	if (xdp_no_room(ipv4_hdr + 1, data_end))
+		return XDP_DROP;
+
+#ifdef CIDR4_FILTER
+	__builtin_memcpy(pfx.lpm.data, &ipv4_hdr->saddr, sizeof(pfx.addr));
+	pfx.lpm.prefixlen = 32;
+
+#ifdef CIDR4_LPM_PREFILTER
+	if (map_lookup_elem(&CIDR4_LMAP_NAME, &pfx))
+		return XDP_DROP;
+	else
+#endif /* CIDR4_LPM_PREFILTER */
+		return map_lookup_elem(&CIDR4_HMAP_NAME, &pfx) ?
+		       XDP_DROP : check_v4_endpoint(xdp, ipv4_hdr);
+#else
+	return check_v4_endpoint(xdp, ipv4_hdr);
```

Comment:
Why is the CIDR filter and the endpoint lookup mutually exclusive?

Reply:
Why mutually exclusive? In the above, when CIDR filter is installed and we match, then we drop, but if CIDR filter passes, then we continue on the ep lookup. If compiled out, we only do the ep lookup. Can you elaborate?

Reply:
Sorry, missed the 2nd call to `check_v4_endpoint`. Initially, I only saw the one in the `#else` branch and got confused.